stacks-signer/src/signerdb.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::path::PathBuf;
+
+use blockstack_lib::util_lib::db::{query_row, table_exists, tx_begin_immediate, Error as DBError};
+use r2d2::PooledConnection;
+use r2d2_sqlite::SqliteConnectionManager;
+use rusqlite::{Connection, Error as SqliteError, ToSql, Transaction as SqlTransaction, NO_PARAMS};
+use stacks_common::util::hash::Sha512Trunc256Sum;
+
+use crate::runloop::BlockInfo;
+
+/// This struct manages a SQLite database connection
+/// for the signer.
+#[derive(Debug, Clone)]
+pub struct SignerDb {
+    /// The SQLite database path
+    pub db_path: PathBuf,
+    /// Connection pool. A pool is used to allow for
+    /// thread-safe connections
+    pub pool: r2d2::Pool<SqliteConnectionManager>,
+}
+
+const CREATE_TABLE: &'static str = "
+CREATE TABLE IF NOT EXISTS events (
+    id INTEGER PRIMARY KEY,
+    topic TEXT NOT NULL
+)";
+
+const CREATE_BLOCKS_TABLE: &'static str = "
+CREATE TABLE IF NOT EXISTS blocks (
+    signer_signature_hash TEXT PRIMARY KEY,
+    block_info TEXT NOT NULL
+)";
+
+impl SignerDb {
+    /// Create a new `SignerState` instance.
+    /// This will create a new SQLite database at the given path
+    /// if one doesn't exist.
+    pub fn new(db_path: &PathBuf) -> Result<SignerDb, DBError> {
```

Comment:
Just a nit. I would maybe make this an optional. If it is none, then open in mem :) instead of enforcing a specific string.